File: Week3/Code/Vectorize1.R
#!/usr/bin/env Rscript
# Author: David Scott
# Contact: david.scott18@imperial.ac.uk
# Date:  October 24 2018
# Description: sums all elements of a matrix 

rm(list=ls()) # clears workspace
graphics.off() # clears graphics

# creates matix of uniform distribution 
M <- matrix(runif(1000000),1000,1000)

SumAllElements <- function(M){
  Dimensions <- dim(M)
  Tot <- 0
  for (i in 1:Dimensions[1]){
    for (j in 1:Dimensions[2]){
      Tot <- Tot + M[i,j]
    }
  }
  return (Tot)
}

## This on my computer takes about 1 sec
print("Speed of SumAllElements function defined using loops:")
print(system.time(SumAllElements(M)))

## While this takes about 0.01 sec
print("Speed of sum function without loops:")
print(system.time(sum(M)))

#system.time function calculates time taken by code
#sum() function 100 times faster than SumAllElements()
# as it uses vectorisation, avoiding the amount of loops used by SumAllElements

File: Week3/Code/Vectorize1.py
#!/usr/bin/env python3
# Date: October 2018

""" Compares speed of two functions with and without vectorization""" 

__appname__ = '[Vectorize1.py]'
__author__ = 'David Scott (david.scott18@imperial.ac.uk)'
__version__ = '0.0.1'
__license__ = "License for this code/program"

## imports ##
import numpy as np 
import time

# creates random array of dimension 1000 * 10000
#   with uniform distribution
M = np.random.rand(1000, 1000)

# define function to count each element of array M
def SumAllElements(M):
    """Counts sum of elements of an array """
    tot = 0
    for i in np.nditer(M): 
        tot += i
    return tot

## Call SumAllElements function with input and timeit 
start1 = time.time()  # starts timer
SumAllElements(M) #calls function with input
end1 = time.time()  # ends timer

## Call sum function with input and timeit 
# sum function does that same as above but without loops
start2 = time.time() # starts timer
np.sum(M)
end2 = time.time() # ends timer

# print speed of functiosn to compare
print("Speed of SumAllElements function using loops: {}".format(end1 - start1)) 
print("Speed of sum function without loops: {}".format(end2 - start2)) 
